src/server/pfs/server/driver.go

```diff
@@ -488,6 +483,62 @@ func (d *driver) deleteProject(txnCtx *txncontext.TransactionContext, project *p
 	return nil
 }
 
+// Set child.ParentCommit (if 'parent' has been determined) and write child to parent's ChildCommits
+func (d *driver) linkParent(txnCtx *txncontext.TransactionContext, child *pfs.CommitInfo, parent *pfs.Commit, needsFinishedParent bool) error {
+	if parent == nil {
+		return nil
+	}
+	// Resolve 'parent' if it's a branch that isn't 'branch' (which can
+	// happen if 'branch' is new and diverges from the existing branch in
+	// 'parent').
+	// Clone the parent proto because resolveCommit will modify it.
+	parent = proto.Clone(parent).(*pfs.Commit)
+	parentCommitInfo, err := d.resolveCommit(txnCtx.SqlTx, parent)
+	if err != nil {
+		return errors.Wrapf(err, "parent commit not found")
+	}
+	// fail if the parent commit has not been finished
+	if needsFinishedParent && parentCommitInfo.Finishing == nil {
+		return errors.Errorf("parent commit %s has not been finished", parentCommitInfo.Commit)
+	}
+	child.ParentCommit = parentCommitInfo.Commit
+	parentCommitInfo.ChildCommits = append(parentCommitInfo.ChildCommits, child.Commit)
+	// Note: error is emitted if parent.ID is a missing/invalid branch OR a
+	// missing/invalid commit ID
+	return errors.Wrapf(d.commits.ReadWrite(txnCtx.SqlTx).Put(parentCommitInfo.Commit, parentCommitInfo),
+		"could not resolve parent commit %s", parent)
+}
+
+// TODO(acohen4): reassess need for "needsFinishedParent" parameter
```

Comment:
Undone TODO

Reply:
Requiring source commits to have finishing / finished parents ensures that the commits are not compacted in a pathological order (finishing later commits before earlier commits will result with us compacting the earlier commits multiple times).